ptvsd/wrapper.py

```diff
@@ -750,30 +750,53 @@ def on_variables(self, request, args):
                 var['variablesReference'] = self.var_map.to_vscode(
                     pyd_child, autogen=True)
 
-            if len(pyd_var) > 3:
-                # This means the current variable has a parent i.e, it is not a
-                # FRAME variable. These require evaluateName to work in VS 
-                # watch window
-                eval_name = pyd_var[3]
-                for s in pyd_var[4:]:
-                    try:
-                        # In the result XML from pydevd list indexes appear 
-                        # as names. If the name is a number then it is a index.
-                        eval_name += '[' + str(int(s)) + ']'
-                    except:
-                        eval_name += '.' + s 
-
-                try:
-                    eval_name += '[' + str(int(var_name)) + ']'
-                except:
-                    eval_name += '.' + var_name
-
+            eval_name = self.__get_variable_evaluate_name(pyd_var, var_name)
+            if eval_name:
                 var['evaluateName'] = eval_name
 
             variables.append(var)
 
         self.send_response(request, variables=variables)
 
+    def __get_variable_evaluate_name(self, pyd_var_parent, var_name):
+        # TODO: docstring
+        eval_name = None
+        if len(pyd_var_parent) > 3:
+            # This means the current variable has a parent i.e, it is not a
+            # FRAME variable. These require evaluateName to work in VS 
+            # watch window
+            var = pyd_var_parent + (var_name,)
+            print(var)
+            eval_name = var[3]
+            for s in var[4:]:
+                try:
+                    # check if this is a dictionary key
+                    # Note: this is best effort, keys that are object references
+                    # will not work
+                    key = self.__get_dictionary_key(s)
+                    print(key)
+                    if key is not None:
+                        eval_name += '[{}]'.format(key)
+                    else:
+                        # In the result XML from pydevd list indexes appear 
+                        # as names. If the name is a number then it is a index.
+                        eval_name += '[{}]'.format(int(s))
+                except:
```

Comment:
The `try`..`except` is now catching all errors from `__get_dictionary_key` and others. If that's not intentional, I think it should go just around `int(s)` - I think you could just do `key = int(s)` under `try`, and then the actual string formatting is identical for both branches.

Reply:
I agree, will reduce this.